### PR TITLE
TCOMP-1884 : Add getValue & with method on Record.Builder

### PR DESCRIPTION
**Pull request by clesaec**
https://jira.talendforge.org/browse/TCOMP-1884

Add util method on Record.Builder
Unit tests added on component-api & component-runtime-impl
Correction done on sanitzeConnectionName on component-api (after fails on unit test)

### Requirements

* Any code change adding any logic **MUST** be tested through a unit test executed with the default build
* Any API addition **MUST** be done with a documentation update if relevant

### CODE REVIEW

File: component-api/src/main/java/org/talend/sdk/component/api/component/AfterVariables.java [hunk 1]
```diff
@@ -33,8 +33,7 @@
  */
 @Documentation("Declare the group of after variable `@AfterVariable`, "
         + "only supported on components - `@PartitionMapper`, `@Processor`, `@Emitter`."
-        + "The functionality is for the Studio only."
-)
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/component/AfterVariables.java:37]: please remove anything not related to PR.

File: component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java [hunk 2]
```diff
@@ -200,21 +231,26 @@ default Entry getEntry(final String name) {
...
-                sanitized[i - offset] = '_';
+            char current = original[i];
+            if ((!ascii.canEncode(current)) || (!Character.isLetterOrDigit(current) && current != '_')) {
+                sanitized[i] = '_';
             } else {
-                sanitized[i - offset] = original[i];
+                sanitized[i] = current;
             }
         }
         return new String(sanitized);
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java:256]: Sorry but please attach this change to a new jira issue. This is really a bug fix and not related to PR but great catch 👍🏻

File: component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java [hunk 3]
```diff
@@ -60,16 +66,41 @@ default Entry getEntry(final String name) {
...
+        /**
+         * Check if input can be affected to an entry of this type.
+         * 
+         * @param input : object.
+         * @return true if input is null or ok.
+         */
+        public boolean isCompatible(final Object input) {
+            if (input == null) {
+                return true;
+            }
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java:96]: just a remark, we have the `nullable` in entry. checking rest of code...

File: component-api/src/main/java/org/talend/sdk/component/api/service/record/RecordBuilderFactory.java [hunk 4]
```diff
@@ -62,4 +65,27 @@
...
      */
     Schema.Entry.Builder newEntryBuilder();
+
+    /**
+     * Build a schema.entry from another one. Useful to duplicate a column with some changes.
+     * 
+     * @param model : model of entry to copy.
+     * @return entry builder with model parameters.
+     */
+    default Schema.Entry.Builder newEntryBuilder(Schema.Entry model) {
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/service/record/RecordBuilderFactory.java:75]: Why not... but don't we have enough?

File: component-api/src/test/java/org/talend/sdk/component/api/service/interceptor/InterceptorHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        final MyAnnot annot = m1.get();
+        Assertions.assertEquals("C1.m1", annot.xx());
+
+        final Optional<MyAnnot> m2 = handler.findAnnotation(C2.class.getDeclaredMethod("m2"), MyAnnot.class);
+        Assertions.assertNotNull(m2);
+        Assertions.assertTrue(m2.isPresent());
+        final MyAnnot annot2 = m2.get();
+        Assertions.assertEquals("C2.m2", annot2.xx());
+    }
+}
```
Comment (undx) [inline component-api/src/test/java/org/talend/sdk/component/api/service/interceptor/InterceptorHandlerTest.java:74]: hum, maybe too tired but I don't get the point of this test class related to PR...

File: component-runtime-beam/src/test/java/org/talend/sdk/component/runtime/beam/spi/record/AvroRecordBuilderTest.java [hunk 6]
```diff
@@ -73,12 +73,12 @@ void copyRecord() {
...
+                .withRecord("_address",
                         factory.newRecordBuilder(address).withString("street", "here").withInt("number", 1).build())
                 .build();
         final Record output = factory.newRecordBuilder(customSchema, baseRecord).withString("custom", "added").build();
         assertEquals(
-                "AvroRecord{delegate={\"name\": \"Test\", \"age\": 33, \"address\": {\"street\": \"here\", \"number\": 1}, \"custom\": \"added\"}}",
+                "AvroRecord{delegate={\"name\": \"Test\", \"age\": 33, \"_address\": {\"street\": \"here\", \"number\": 1}, \"custom\": \"added\"}}",
                 output.toString());
     }
 }
```
Comment (undx) [inline component-runtime-beam/src/test/java/org/talend/sdk/component/runtime/beam/spi/record/AvroRecordBuilderTest.java:84]: see my comment upper on splitting PR responsibility....

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java [hunk 7]
```diff
@@ -115,6 +118,37 @@ public BuilderImpl(final Schema providedSchema) {
             this.providedSchema = providedSchema;
         }
 
+        @Override
+        public Object getValue(final String name) {
+            return this.values.get(name);
+        }
+
+        @Override
+        public Builder with(final Entry entry, final Object value) {
```
Comment (undx) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:127]: Why not... but same remark as above.

**Review by undx — COMMENTED** [review]
I'm not totally sure to understand the goal of this PR...
I don't get the point _Record.Builder instances are currently opaque, you can't see current state (values)._
This is supposed to be when using a builder... so following steps should be enough clear, no?

File: component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java (hunk 3 above)
Comment (clesaec) [reply component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java:96]: Here, it's just to check compatibility with Schema.Type enum (as instanceof), so null is OK

File: component-api/src/main/java/org/talend/sdk/component/api/service/record/RecordBuilderFactory.java (hunk 4 above)
Comment (clesaec) [reply component-api/src/main/java/org/talend/sdk/component/api/service/record/RecordBuilderFactory.java:75]: It's a facility function, similar to 
```java
Schema.Builder newSchemaBuilder(Schema schema);
Record.Builder newRecordBuilder(Schema schema, Record record);
```
and can help on DataPrep processor, where we can add a new Field that is a copy of an existing one but with a different name.

File: component-api/src/test/java/org/talend/sdk/component/api/service/interceptor/InterceptorHandlerTest.java (hunk 5 above)
Comment (clesaec) [reply component-api/src/test/java/org/talend/sdk/component/api/service/interceptor/InterceptorHandlerTest.java:74]: Yes, not related to the PR
I needed to watch how InterceptorHandler worked, add this unit test and as this is the only direct unit test for this class and take few millis to run, i thought it was good to keep it.

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 7 above)
Comment (clesaec) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:127]: May simplify client code like : "if (entry.getType() ... && value instanceof) ..."; only constraint is that type control become dynamic (like for nullable).

File: component-api/src/test/java/org/talend/sdk/component/api/service/interceptor/InterceptorHandlerTest.java (hunk 5 above)
Comment (undx) [reply component-api/src/test/java/org/talend/sdk/component/api/service/interceptor/InterceptorHandlerTest.java:74]: ok, let's keep it although not related...

File: component-api/src/test/java/org/talend/sdk/component/api/record/SchemaTest.java [hunk 8]
```diff
@@ -135,6 +135,20 @@ public String getProp(String property) {
...
+        Assertions.assertTrue(Schema.Type.INT.isCompatible(123));
+        Assertions.assertFalse(Schema.Type.INT.isCompatible(234L));
+        Assertions.assertFalse(Schema.Type.INT.isCompatible("Hello"));
+
+        Assertions.assertTrue(Type.BYTES.isCompatible("Hello".getBytes()));
+        Assertions.assertTrue(Type.BYTES.isCompatible(new Byte[] {}));
+        Assertions.assertFalse(Schema.Type.BYTES.isCompatible(new int[] {}));
+
+        Assertions.assertTrue(Type.DATETIME.isCompatible(System.currentTimeMillis()));
+    }
```
Comment (undx) [inline component-api/src/test/java/org/talend/sdk/component/api/record/SchemaTest.java:150]: extend a bit test to all types....

File: component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java [hunk 9]
```diff
@@ -63,16 +66,41 @@ default Entry getEntry(final String name) {
...
-        DATETIME
+        RECORD(new Class<?>[] { Record.class }),
+        ARRAY(new Class<?>[] { Collection.class }),
+        STRING(new Class<?>[] { String.class }),
+        BYTES(new Class<?>[] { byte[].class, Byte[].class }),
+        INT(new Class<?>[] { Integer.class }),
+        LONG(new Class<?>[] { Long.class }),
+        FLOAT(new Class<?>[] { Float.class }),
+        DOUBLE(new Class<?>[] { Double.class }),
+        BOOLEAN(new Class<?>[] { Boolean.class }),
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java:77]: Shouldn't we include primitives classes as in MappingUtils? (ie int.class, etc.)

**Review by undx — APPROVED** [review]
Some minor comments but LGTM.

File: component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java (hunk 9 above)
Comment (clesaec) [reply component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java:77]: As "public boolean isCompatible(final Object input)" as an Object in input, a call with a primitive will be "boxed" into corresponding object; so primitives type is not useful.

File: component-api/src/test/java/org/talend/sdk/component/api/record/SchemaTest.java (hunk 8 above)
Comment (clesaec) [reply component-api/src/test/java/org/talend/sdk/component/api/record/SchemaTest.java:188]: done

**Pull request merged**